apps/aechannel/src/aesc_fsm_min_depth_watcher.erl

```diff
@@ -104,122 +233,707 @@ terminate(_Reason, _St) ->
+            end
+    end.
+
+stop_cond({all_after_tx, TxHash}, _PrevHash, Found) ->
+    case Found of
+        [] -> false;
+        [_|_] ->
+            tail_after_tx(TxHash, Found)
     end;
-watch_for_channel_closing(_, R, _, C) ->
-    {R, C}.
-
-
-channel_status(ChId, C) ->
-    channel_status(maps:find({status, ChId}, C), ChId, C).
-
-channel_status({ok, S}, _, C) -> {S, C};
-channel_status(error, ChId, C) ->
-    St = case aec_chain:get_channel(ChId) of
-             {ok, Ch} ->
-                 #{ is_active    => aesc_channels:is_active(Ch)
-                  , lock_period  => aesc_channels:lock_period(Ch)
-                  , locked_until => aesc_channels:locked_until(Ch) };
-             _ ->
-                 undefined
-         end,
-    {St, C#{ {status, ChId} => St }}.
-
-min_depth_achieved(TxHash, MinDepth, C) ->
+stop_cond({any_after_block, Hash}, PrevHash, Found) ->
+    if Hash == PrevHash ->
+            {true, Found};
+       true ->
+            has_create_tx(Found)
+    end.
+
+%% Safety. If a create_tx exists in Found, it must be the first one.
+has_create_tx([]) ->
+    false;
+has_create_tx([SignedTx|_] = Found) ->
+    case aetx:specialize_type(aetx_sign:tx(SignedTx)) of
+        {channel_create_tx, _} ->
+            {true, Found};
+        _ ->
+            false
+    end.
+
+tail_after_tx(_, []) ->
+    false;
+tail_after_tx(Hash, [H|T]) ->
+    case aetx_sign:hash(H) of
+        Hash ->
+            {true, T};
+        _ ->
+            tail_after_tx(Hash, T)
+    end.
+
+tx_hashes(BlockHash, BlockHdr, C) ->
+    cached_get({tx_hashes, BlockHash}, C,
+               fun() -> tx_hashes_(BlockHash, BlockHdr) end).
+
+tx_hashes_(Hash, Hdr) ->
+    case aec_headers:type(Hdr) of
+        key -> [];
+        micro ->
+            case aec_db:find_block_tx_hashes(Hash) of
+                {value, TxHashes} ->
+                    TxHashes;
+                not_found ->
+                    []
+            end
+    end.
+
+txs_for_chid(ChId, Hash, Hdr, C) ->
+    {TxHashes, C1} = tx_hashes(Hash, Hdr, C),
+    lists:foldl(
+      fun(TxHash, {Acc, Cx}) ->
+              {SignedTx, Cx1} = get_signed_tx(TxHash, Cx),
+              lager:debug("SignedTx = ~p", [SignedTx]),
+              case is_tx_for_chid(SignedTx, ChId) of
+                  true  -> {[SignedTx|Acc], Cx1};
+                  false -> {Acc, Cx1}
+              end
+      end, {[], C1}, TxHashes).
+
+get_signed_tx(TxHash, C) ->
+    cached_get({signed_tx, TxHash}, C, fun() -> get_signed_tx_(TxHash) end).
+
+get_signed_tx_(TxHash) ->
+    case aec_db:find_signed_tx(TxHash) of
+        {value, STx} ->
+            STx;
+        none ->
+            undefined
+    end.
+
+
+sort_txs(Txs) ->
+    lists:sort(fun compare_txs/2, Txs).
+
+%% See lists:sort/2: The fun Comp(A,B) should return true if A =< B
+compare_txs(SignedA, SignedB) ->
+    aetx:nonce(aetx_sign:tx(SignedA)) =< aetx:nonce(aetx_sign:tx(SignedB)).
```

Comment:
Unclear to me how this will work for:
* Txs from different senders (initiator / responder). I do not have full understanding of assumptions of code on this ordering.
* Generic account meta tx' nonce - that I consider out of scope for this PR cc @velzevur 

I see no issues so far though.

Reply:
I see changed in f9afc7b41c8e69f72edd677e26823135461611b4